File: solutions/csharp/secret-handshake/1/SecretHandshake.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class SecretHandshake
{
	public static string[] Commands(int commandValue)
	{
		Signal command = (Signal)commandValue;
		List<Signal> signals = command.GetSignals()
			.ToList();

		if (signals.Contains(Signal.Reverse))
		{
			_ = signals.Remove(Signal.Reverse);
			signals.Reverse();
		}

		return signals.Select(x => x.ToValue())
			.ToArray();
	}

	private static IEnumerable<Signal> GetSignals(this Signal values)
	{
		foreach (Signal signal in Enum.GetValues(typeof(Signal)))
		{
			if (values.HasFlag(signal))
			{
				yield return signal;
			}
		}
	}

	private static string ToValue(this Signal signal)
	{
		SignalTextAttribute[] attributes = GetAttributes(signal);
		return attributes.Length > 0
			? attributes[0].Value.ToString()
			: "";
	}

	private static SignalTextAttribute[] GetAttributes(this Signal signal)
	{
		SignalTextAttribute[] attributes = (SignalTextAttribute[])signal
		   .GetType()
		   .GetField(signal.ToString())
		   .GetCustomAttributes(typeof(SignalTextAttribute), false);

		return attributes;
	}
}

[Flags]
public enum Signal
{
	[SignalText("wink")]
	Wink = 1,

	[SignalText("double blink")]
	DoubleBlink = 2,

	[SignalText("close your eyes")]
	CloseYourEyes = 4,

	[SignalText("jump")]
	Jump = 8,

	Reverse = 16,
}

internal class SignalTextAttribute : ValueAttribute
{
	public SignalTextAttribute(string value) : base(value)
	{
	}
}

internal abstract class ValueAttribute : Attribute
{
	public ValueAttribute(string value) =>
	Value = value;

	public string Value { get; }
}

File: solutions/scala/secret-handshake/1/src/main/scala/SecretHandshake.scala
case class SecretHandshake()

object SecretHandshake {
  def commands(value: Int): List[String] =
    commands(binaryComponents(value))

  private def commands(binaryComponents: List[Int]): List[String] = {
    var signals = Signal.fromValues(binaryComponents)
    if (signals.contains(Signal.Reverse)) {
      signals = signals.filter(signal => signal != Signal.Reverse)
                       .reverse
    }

    signals.map(signal => signal.text)
  }

  private def binaryComponents(n: Int): List[Int] = {
    if (n == 0) {
      Nil
    }
    else if ((n & 1) == 1) {
      1 :: binaryComponents(n >> 1).map(_ * 2)
    }
         else {
           binaryComponents(n >> 1).map(_ * 2)
         }
  }
}
